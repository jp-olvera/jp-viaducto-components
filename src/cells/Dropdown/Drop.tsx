import React, { useRef, useEffect } from 'react';
import { createPortal } from 'react-dom';
import { findScrollParents } from '../../utils/scroll';

interface DropProps {
  target: React.RefObject<HTMLElement> | any;
  contentRef: React.RefObject<HTMLElement> | any;
  children: React.ReactNode;
}
const Drop = ({ target, children, contentRef }: DropProps) => {
  const dropRef = useRef<HTMLDivElement>(null);

  const move = () => {
    if (contentRef.current && target.current && dropRef.current) {
      const windowHeight = window.innerHeight;
      const windowWidth = document.body.clientWidth;
      const targetRect: DOMRect = target.current.getBoundingClientRect();
      const dropH = contentRef.current?.offsetHeight || 0; // height drop
      const dropW = contentRef.current?.offsetWidth || 0; // height drop
      let top = targetRect.bottom;
      const dropR = targetRect.left + dropW;
      // ajustar verticalmente
      if (windowHeight - targetRect.bottom > dropH) {
        // abajo hay espacio
        top = targetRect.bottom + 3;
      } else if (targetRect.top >= dropH) {
        // arriba cabe
        top = targetRect.top - dropH - 3;
      } else {
        top = targetRect.bottom + 3;
      }
      let { left } = targetRect;

      // ajustar a los lados
      if (windowWidth > dropW) {
        // la ventana es mayor que el drop
        if (dropR > windowWidth) {
          // el drop se sale a la derecha
          left = targetRect.right - dropW;
        } else {
          // el drop no se sale a la derecha
          left = targetRect.left;
        }
      } else {
        // la ventana es más pequeña
        left = 0;
      }
      dropRef.current.style.top = `${top}px`;
      dropRef.current.style.left = `${left}px`;
    }
  };

  useEffect(() => {
    move();
    let scrollParents: (Element | Document)[] = [];
    const addScrollListeners = () => {
      scrollParents = findScrollParents(target.current);
      scrollParents.forEach((scrollParent) => scrollParent.addEventListener('scroll', move));
    };
    const removeScrollListeners = () => {
      scrollParents.forEach((scrollParent) => scrollParent.removeEventListener('scroll', move));
      scrollParents = [];
    };
    if (target && target.current) {
      addScrollListeners();
      window.addEventListener('resize', move);
    }
    return function cleanup() {
      removeScrollListeners();
      window.removeEventListener('resize', move);
    };
  }, []);

  useEffect(() => {
    if (dropRef.current) {
      dropRef.current.focus();
      if (target.current?.offsetWidth) {
        dropRef.current.style.minWidth = `${target.current.offsetWidth}px`;
      }
    }
  }, [dropRef, target]);
  /* istanbul ignore if */
  if (!target || !target.current) {
    return null;
  }
  return createPortal(
    <div
      ref={dropRef}
      style={{
        position: 'fixed',
        zIndex: 1,
      }}
      tabIndex={0}
      role='listbox'
    >
      {children}
    </div>,
    document.body,
  );
};
export default Drop;
