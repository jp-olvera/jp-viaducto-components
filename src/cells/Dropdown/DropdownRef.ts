/* istanbul ignore file */

export const refs = {
  clickOutsideHandler: (
    event: any,
    activatorRef: any,
    dropdownListRef: any,
    setIsOpen: Function,
  ) => {
    if (dropdownListRef.current && activatorRef.current) {
      if (
        dropdownListRef.current.contains(event.target)
        || activatorRef.current.contains(event.target)
      ) {
        return;
      }
    }
    setIsOpen(false);
  },
  clickHandler: (
    setIsOpen: Function,
    isOpen: boolean,
    dropdownListRef: any,
    wrapperRef: any,
  ) => {
    setIsOpen(!isOpen);
    if (dropdownListRef && dropdownListRef.current && wrapperRef.current) {
      const bounding = dropdownListRef.current.getBoundingClientRect();
      const bottom = wrapperRef.current.clientHeight || '2.4rem';
      if (
        bounding.bottom
        > (window.innerHeight || document.documentElement.clientHeight)
      ) {
        // eslint-disable-next-line no-param-reassign
        dropdownListRef.current.style.bottom = `calc(${bottom}px + 0.5rem`;
      }
    }
  },
};
