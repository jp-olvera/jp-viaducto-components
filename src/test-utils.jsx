import React from 'react';
import { render } from '@testing-library/react';
import { ConfigProvider } from './providers';

const AllTheProviders = ({ children }) => {
  return <ConfigProvider>{children}</ConfigProvider>;
};

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
