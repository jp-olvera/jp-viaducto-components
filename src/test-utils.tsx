import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { ConfigProvider } from './providers';

const AllTheProviders = ({ children }: any) => (
  <ConfigProvider>{children}</ConfigProvider>
);

jest.mock('./cells/Input/amex.svg', () => null);
jest.mock('./cells/Input/mastercard.svg', () => null);
jest.mock('./cells/Input/visa.svg', () => null);
jest.mock('./cells/Dropdown/sorting.svg', () => null);
jest.mock('./cells/BareButton/close.svg', () => null);

const customRender = (ui, options?) => render(ui, { wrapper: AllTheProviders, ...options });
// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
