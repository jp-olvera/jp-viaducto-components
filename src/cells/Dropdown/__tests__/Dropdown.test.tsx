/* eslint-env jest */

import React from 'react';
import { render, fireEvent } from '../../../test-utils';
import '@testing-library/jest-dom/extend-expect';
import { Dropdown } from '..';
import Drop from '../Drop';
import { refs } from '../DropdownRef';

const options = [<p>1</p>, <p>2</p>, <p>3</p>];

describe('<Dropdown/>', () => {
  test('should render an focusable activator to toggle the dropdown', () => {
    const dropdown = render(
      <Dropdown
        content={options}
        family={null}
        defaultText=''
        onClick={() => {}}
      />,
    );
    expect(dropdown.getByTestId('dropdown-activator')).toBeInTheDocument();
    const activator = dropdown.getByTestId('dropdown-activator');
    activator.focus();
    expect(activator).toHaveFocus();
    fireEvent.mouseUp(activator);
  });
  test('should display popup', () => {
    const mockFn = jest.fn();
    const dropdown = render(
      <Dropdown
        content={options}
        family='Roboto'
        onClick={mockFn}
        defaultText=''
      />,
    );
    const activator = dropdown.getByTestId('dropdown-activator');
    activator.click();
    expect(dropdown.getByTestId('dropdown-itemList')).toBeVisible();
    fireEvent.click(dropdown.queryByText('1') || window);
    expect(mockFn).toHaveBeenCalled();
  });
  test('should render with default props and borders', () => {
    const borders = {
      top: '1px solid black',
      bottom: '1px solid black',
      left: '1px solid black',
      right: '1px solid black',
    };
    const { container } = render(
      <Dropdown border={borders} defaultText='' onClick={() => {}} />,
    );
    expect(container).toMatchSnapshot();
  });
  test('should render with default props and border left none', () => {
    const bordersNoLeft = {
      top: '1px solid black',
      bottom: '1px solid black',
      right: '1px solid black',
    };
    const { container } = render(
      <Dropdown border={bordersNoLeft} onClick={() => {}} defaultText='' />,
    );
    expect(container).toMatchSnapshot();
  });
  test('should click outside of dropdrown', () => {
    const { container, queryByText } = render(
      <section>
        <Dropdown
          content={options}
          family='Roboto'
          onClick={() => {}}
          defaultText='Selection'
        />
      </section>,
    );
    const activator = queryByText('Selection');
    fireEvent.click(activator);
    const section = container.querySelector('section');
    fireEvent.mouseUp(section);
    fireEvent.click(section);
    expect(container).toMatchSnapshot();
  });

  describe('<Drop>', () => {
    const ref = React.createRef();
    test('should render Drop', () => {
      const { container } = render(
        <Drop target={<div />} contentRef={ref}>
          <div>aa</div>
        </Drop>,
      );
      expect(container).toMatchSnapshot();
    });
  });
  describe('ref functions', () => {
    test('should not be null in clickHandler', () => {
      jest.spyOn(refs, 'clickHandler').mockImplementation(() => 1);
      expect(
        refs.clickHandler(
          () => {},
          true,
          {
            current: {
              getBoundingClientRect: () => ({ bottom: 500 }),
              style: { bottom: '' },
            },
          },
          { current: { clientHeight: '1rem' } },
        ),
      ).not.toBeNull();
    });
    test('should not be null in clickOutsideHandler', () => {
      jest.spyOn(refs, 'clickOutsideHandler').mockImplementation(() => null);
      expect(refs.clickOutsideHandler).not.toBeNull();
    });
  });
});
