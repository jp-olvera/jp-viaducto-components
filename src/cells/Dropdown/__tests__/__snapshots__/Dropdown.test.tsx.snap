// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dropdown/> <Drop> should render Drop 1`] = `<div />`;

exports[`<Dropdown/> <Drop> should render Drop 2`] = `
<div>
  <div
    style="height: 90rem; width: 750rem;"
  />
</div>
`;

exports[`<Dropdown/> should click outside of dropdrown 1`] = `
<div>
  <section>
    <div
      class="StyledDropdown__Wrapper-sc-1btfu08-0 dQNbzS"
      height="2.488rem"
    >
      <button
        aria-controls="dropdown1"
        aria-haspopup="true"
        aria-selected="true"
        class="StyledDropdown__Activator-sc-1btfu08-1 hAeXvx"
        data-testid="dropdown-activator"
        id="dropdown-activator"
        type="button"
      >
        <span>
          Selection
        </span>
        <img
          alt=""
        />
      </button>
    </div>
  </section>
</div>
`;

exports[`<Dropdown/> should render with default props and border left none 1`] = `
<div>
  <div
    class="StyledDropdown__Wrapper-sc-1btfu08-0 dQNbzS"
    height="2.488rem"
  >
    <button
      aria-controls="dropdown1"
      aria-haspopup="true"
      aria-selected="true"
      class="StyledDropdown__Activator-sc-1btfu08-1 bONsQJ"
      data-testid="dropdown-activator"
      id="dropdown-activator"
      type="button"
    >
      <span />
      <img
        alt=""
      />
    </button>
  </div>
</div>
`;

exports[`<Dropdown/> should render with default props and borders 1`] = `
<div>
  <div
    class="StyledDropdown__Wrapper-sc-1btfu08-0 dQNbzS"
    height="2.488rem"
  >
    <button
      aria-controls="dropdown1"
      aria-haspopup="true"
      aria-selected="true"
      class="StyledDropdown__Activator-sc-1btfu08-1 bONsQJ"
      data-testid="dropdown-activator"
      id="dropdown-activator"
      type="button"
    >
      <span />
      <img
        alt=""
      />
    </button>
  </div>
</div>
`;
